src/daft-distributed/src/scheduling/dispatcher.rs

```diff
@@ -0,0 +1,136 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use common_error::{DaftError, DaftResult};
+use common_partitioning::PartitionRef;
+use futures::FutureExt;
+use tokio_util::sync::CancellationToken;
+
+use crate::{
+    channel::{
+        create_channel, create_oneshot_channel, OneshotReceiver, OneshotSender, Receiver, Sender,
+    },
+    runtime::JoinSet,
+    scheduling::{task::SwordfishTask, worker::WorkerManager},
+};
+// The task dispatcher is responsible for dispatching tasks to workers.
+#[allow(dead_code)]
+pub(crate) struct TaskDispatcher {
+    worker_manager: Box<dyn WorkerManager>,
+}
+
+impl TaskDispatcher {
+    pub fn new(worker_manager: Box<dyn WorkerManager>) -> Self {
+        Self { worker_manager }
+    }
+
+    pub fn spawn_task_dispatcher(
+        task_dispatcher: Self,
+        joinset: &mut JoinSet<DaftResult<()>>,
+    ) -> TaskDispatcherHandle {
+        let (task_dispatcher_sender, task_dispatcher_receiver) = create_channel(1);
+        joinset.spawn(Self::run_dispatch_loop(
+            task_dispatcher,
+            task_dispatcher_receiver,
+        ));
+        TaskDispatcherHandle::new(task_dispatcher_sender)
+    }
+
+    async fn run_dispatch_loop(
+        _dispatcher: Self,
+        _task_rx: Receiver<DispatchableTask>,
+    ) -> DaftResult<()> {
+        todo!("Implement run dispatch loop");
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct TaskDispatcherHandle {
+    task_dispatcher_sender: Sender<DispatchableTask>,
+}
+
+impl TaskDispatcherHandle {
+    fn new(task_dispatcher_sender: Sender<DispatchableTask>) -> Self {
+        Self {
+            task_dispatcher_sender,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn submit_task(&self, task: SwordfishTask) -> DaftResult<SubmittedTask> {
+        let (result_tx, result_rx) = create_oneshot_channel();
+        let cancel_token = CancellationToken::new();
+        let dispatchable_task = DispatchableTask::new(task, result_tx, cancel_token.clone());
```

Comment:
you can like clean this up by having an api like:

```
let dispatchable_task = DispatchableTask::new(result_tx);
let submitted_task = dispatchable_task.dispatch(self.task_dispatcher_sender).await;
```

we can then abstract away the CancellationToken and senders